### PR TITLE
[FIX #133] Various exception handler fixes for stability

### LINKED ISSUE #133: [BUG] C-d from remote prompt after connection closed causes uncaught exception
## Describe the bug
If you are connected to a victim, and the remote shell is killed, if you press `C-d` to exit the remote shell, an uncaught `ChannelError` is raised. The routine in `Manager.interactive` needs to be modified to catch these exceptions.

In most places, exceptions are caught, but there was some wiggle-room left during transition between states, which causes this problem.

## Describe the target system
Any target will trigger this error if the remote shell/process is killed and the next input from the attacker is `C-d`. 

## Steps To Reproduce
1. **On victim**: `socat TCP-LISTEN:4444,reuseaddr,fork EXEC:/bin/bash`
2. **On Attacker**: `pwncat W.X.Y.Z 4444`
3. **On Attacker**: Press `C-d` to go to remote prompt
4. **On victim**: Press `C-c` to kill `socat`
5. **On Attacker**: Press `C-d` to return to pwncat prompt

This will cause the uncaught `ChannelError`, and exit pwncat completely. It's worth noting that if any other keys are pressed besides `C-d`, pwncat will recognize the channel has closed and continue correctly.

## Expected behavior
Exception is caught just as if you pressed any other key.

## Screenshots
image

## Expected Solution
The problem is at lines 593-612 of `manager.py`:

```py
            try:
                self.target.platform.interactive_loop(interactive_complete)
            except RawModeExit:
                pass
            except ChannelClosed:
                channel_closed = True
                self.log(
                    f"[yellow]warning[/yellow]: {self.target.platform}: connection reset"
                )
            except Exception:
                pwncat.util.console.print_exception()

            # Trigger thread to exit
            interactive_complete.set()
            output_thread.join()

            # Exit interactive mode
            if channel_closed:
                self.target.died()
            else:
                self.target.platform.interactive = False
```

The `Exit interactive mode` block should be within the `try`-block. Something like this:

```py
            try:
                try:
                    self.target.platform.interactive_loop(interactive_complete)
                except RawModeExit:
                    pass
                self.target.platform.interactive = False
            except ChannelClosed:
                self.target.died()
                self.log(
                    f"[yellow]warning[/yellow]: {self.target.platform}: connection reset"
                )
            except Exception:
                pwncat.util.console.print_exception()

            # Trigger thread to exit
            interactive_complete.set()
            output_thread.join()
```

### PR DESCRIPTION
**Pull request by calebstewart**
## Description of Changes
Fixes #133.

Added better exception handling in `Manager.interactive` which should now withstand dropped sessions. 

## Major Changes Implemented:
- Improved exception handling in across the framework.

## Pre-Merge Tasks
- [x] Formatted all modified files w/ `python-black`
- [x] Sorted imports for modified files w/ `isort`
- [x] Ran `flake8` on repo, and fixed any new problems w/ modified files
- [x] Ran `pytest` test cases
- [x] Added brief summary of updates to CHANGELOG (under `[Unreleased]`)

**For issues with pre-merge tasks, see CONTRIBUTING.md**

### CODE REVIEW

**Comment by calebstewart** [comment]
@Mitul16 if you're interested, want to give this a go and see if it fixes the uncaught exceptions you were seeing before? It seems to fix the ones I could reproduce, but I'm curious to see from your perspective. Thanks!

**Comment by Mitul16** [comment]
It doesn't look like the _fix_ is in action, please do let me know. I would like to test it for sure.

**Comment by calebstewart** [comment]
Sorry for the back and forth. There was a couple cases I missed from your report in the issue. I believe the last push covers all the bases, and I'd really appreciate your feedback now :smile:

**Comment by Mitul16** [comment]
I think the following issue comes under __current Issue__ - `exception handling...`

If I remember correctly, there was once a check in the module `upload.py` checking for _write access_ to the _destination path_
If we try to upload a file to a destination to which the current `session.current_user()` doesn't have access to, then `pwncat` hangs and we have to kill the current _instance_

> pwncat  v0.3.1 - `./pwncat/commands/upload.py`
```python
60             access = pwncat.victim.access(args.destination)
61             if Access.DIRECTORY in access:
62                 args.destination = os.path.join(
63                     args.destination, os.path.basename(args.source)
64                 )
65             elif Access.PARENT_EXIST not in access:
66                 console.log(
67                     f"[cyan]{args.destination}[/cyan]: no such file or directory"
68                 )
69                 return
```
> NOTE: the file to be uploaded should be more than the buffer size (2048 or 4096, or whatever is in use)

**Comment by calebstewart** [comment]
> I think the following issue comes under **current Issue** - `exception handling...`
> 
> If I remember correctly, there was once a check in the module `upload.py` checking for _write access_ to the _destination path_
> If we try to upload a file to a destination to which the current `session.current_user()` doesn't have access to, then `pwncat` hangs and we have to kill the current _instance_
> 
> > pwncat  v0.3.1 - `./pwncat/commands/upload.py`
> 
> ```python
> 60             access = pwncat.victim.access(args.destination)
> 61             if Access.DIRECTORY in access:
> 62                 args.destination = os.path.join(
> 63                     args.destination, os.path.basename(args.source)
> 64                 )
> 65             elif Access.PARENT_EXIST not in access:
> 66                 console.log(
> 67                     f"[cyan]{args.destination}[/cyan]: no such file or directory"
> 68                 )
> 69                 return
> ```
> 
> > NOTE: the file to be uploaded should be more than the buffer size (2048 or 4096, or whatever is in use)

Yes, that used to be a thing, and was removed because I thought the `Popen` wrapper would catch the failed process, but I was obviously wrong. I've added back in the permission checks in `Linux.open` to ensure the files are writable. I also needed to implement a custom `LinuxPath` class to handle `readable()` and `writable()` differently, since we can't enumerate users/groups, because those enumerations open files which causes an infinite loop. Instead, we use cached results from the `id` command to identify privileges to reading/writing.

**Comment by Mitul16** [comment]
Well, now it is breezing :grinning: 

I have tested the following for an invalidated session :
- Switching to and from the remote terminal
- Up/Downloading a file
- Running an `enumerate`, `escalate` module

None of the above renders the `pwncat` instance to exit abruptly
Although I did manage to fix some exceptions myself :sweat_smile: 
But you did it much better, thanks a lot.

There is one little issue, I tried to upload `rockyou.txt` file and then closed the session from the `victim` machine
The good part is that `pwncat` didn't exit, but the _newly invalidated sessions_ are still present and do not dispose when trying `Ctrl-D` to go to the remote terminal instead we get `channel not connected` and some _traceback_

**Comment by calebstewart** [comment]
> Well, now it is breezing 
> 
> I have tested the following for an invalidated session :
> 
> * Switching to and from the remote terminal
> * Up/Downloading a file
> * Running an `enumerate`, `escalate` module
> 
> None of the above renders the `pwncat` instance to exit abruptly
> Although I did manage to fix some exceptions myself 
> But you did it much better, thanks a lot.
> 
> There is one little issue, I tried to upload `rockyou.txt` file and then closed the session from the `victim` machine
> The good part is that `pwncat` didn't exit, but the _newly invalidated sessions_ are still present and do not dispose when trying `Ctrl-D` to go to the remote terminal instead we get `channel not connected` and some _traceback_

Cool, I was able to replicate that pretty easily. The problem was that the `finally:` block was trying to close the Popen object, which caused a chain of `ChannelClosed -> ChannelError`. The ChannelError was caused by calling `send` with a closed channel. Instead, I changed the base `Socket` class to raise another `ChannelClosed` exception instead of a generic `ChannelError` in that case. This makes it so that the error propagates properly and closes the session.

image

**Comment by Mitul16** [comment]
@calebstewart I have a doubt, why have you used `ChannelError(ERROR_MESSAGE)` in some places instead of `ChannelError(CHANNEL, ERROR_MESSAGE)`?

In `./pwncat/channel/__init__.py`, we have the `ChannelError` constructor like so
```python
 35     def __init__(self, ch, msg="generic channel failure"):
 36         super().__init__(msg)
 37         self.channel = ch
```

Even in the _classes_ extending the _super class_ `ChannelError`, the _constructor_ for `ChannelError` is called as
```python
def __init__(self, ch):
    super().__init__(ch, CORRESPONDING_ERROR_MESSAGE)
```

The places where you have `ChannelError(ERROR_MESSAGE)` will not raise any _exceptions_ as there is a _default argument_ in its _constructor_ because of this any `ERROR_MESSAGE` that you use (without passing a `CHANNEL`) is simply not shown, __generic channel failure__ is shown instead.

**Comment by Mitul16** [comment]
Regarding my last comment, I have added a commit to the branch [issue-133-uncaught-channelerror](https://github.com/calebstewart/pwncat/tree/issue-133-uncaught-channelerror)
If you haven't checked these issues yourself, I can create a _PR_

**Comment by calebstewart** [comment]
> @calebstewart I have a doubt, why have you used `ChannelError(ERROR_MESSAGE)` in some places instead of `ChannelError(CHANNEL, ERROR_MESSAGE)`?
> 
> In `./pwncat/channel/__init__.py`, we have the `ChannelError` constructor like so
> 
> ```python
>  35     def __init__(self, ch, msg="generic channel failure"):
>  36         super().__init__(msg)
>  37         self.channel = ch
> ```
> 
> Even in the _classes_ extending the _super class_ `ChannelError`, the _constructor_ for `ChannelError` is called as
> 
> ```python
> def __init__(self, ch):
>     super().__init__(ch, CORRESPONDING_ERROR_MESSAGE)
> ```
> 
> The places where you have `ChannelError(ERROR_MESSAGE)` will not raise any _exceptions_ as there is a _default argument_ in its _constructor_ because of this any `ERROR_MESSAGE` that you use (without passing a `CHANNEL`) is simply not shown, **generic channel failure** is shown instead.

You're correct. That appears to be a typo, which would eventually cause an exception as the channel reference is used to cleanup disconnected channels on exception.

**Comment by calebstewart** [comment]
> Regarding my last comment, I have added a commit to the branch [issue-133-uncaught-channelerror](https://github.com/calebstewart/pwncat/tree/issue-133-uncaught-channelerror)
> If you haven't checked these issues yourself, I can create a _PR_

I had not seen it. I went and found it. You can create a pull request if you'd like. I will say I don't agree with catching the ChannelError in `create_session`. These methods aren't front-end calls. They are APIs, and the exception should be propagated to the caller.

**Comment by calebstewart** [comment]
On second thought, I can just merge your changes into my branch. There's no need for a formal pull request. It's all working toward the same goal anyway.

**Comment by Mitul16** [comment]
Ah! You are back. I was testing `ChannelError` specifically and have found some _exceptions_
Here is one example that you can try
```bash
# In `pwncat` terminal prompt, password is `bandit0`
connect ssh://bandit0@bandit.labs.overthewire.org:2220
```

You may find that `getpeername` is not defined for the class `Channel`

**Comment by Mitul16** [comment]
Moreover, I wasn't sure upto how many parent layers (methods) to bring the error, I did find `create_session` a good place to catch the exception

**Comment by calebstewart** [comment]
> Ah! You are back. I was testing `ChannelError` specifically and have found some _exceptions_
> Here is one example that you can try
> 
> ```shell
> # In `pwncat` terminal prompt, password is `bandit0`
> connect ssh://bandit0@bandit.labs.overthewire.org:2220
> ```
> 
> You may find that `getpeername` is not defined for the class `Channel`

That was a bug in the exception handler that slipped through the cracks for a long time. Good catch. Should be fixed in this branch now. Also, I'm pretty sure we're on different time zones :rofl:

> Moreover, I wasn't sure upto how many parent layers (methods) to bring the error, I did find `create_session` a good place to catch the exception

Totally understand your logic, but I would say the `create_session` method is the "entrypoint" call for the API, and should let the exceptions through. It should raise either a `PlatformError` indicating that something is wrong with the platform or a `ChannelError` indicating that something went wrong setting up the channel.

In any case, I'll get your changes merged in a moment.

**Comment by Mitul16** [comment]
This is the reason I didn't create a _PR_ because of the way I have implemented this `exception-handling`
There is no need to merge from my branch, please take your time to implement the same with a _valid approach_

**Comment by calebstewart** [comment]
> This is the reason I didn't create a _PR_ because of the way I have implemented this `exception-handling`
> There is no need to merge from my branch, please take your time to implement the same with a _valid approach_

Not a big deal. I have your fork added as a separate remote to my local clone, so I can easily merge your branch, then make the one slight modification and push back here. I'd prefer that, so that your contributions aren't lost in the git history. :+1:

I've done that, and we should be good at this point at least with the exception problems we've found so far. Hopefully there aren't any more. :sob:

**Comment by calebstewart** [comment]
Are you still seeing any spurious or weird exception issues or working on anything related to this PR? If not, I will merge this branch back into master.

**Comment by Mitul16** [comment]
Unfortunately there are a little more, but these are not related to `ChannelError` or `Channel` in particular
Try `back` in `pwncat` terminal prompt :sweat_smile:

**Comment by calebstewart** [comment]
The back command was super simple fix, so I went ahead and just fixed that. It was raising the wrong exception type.

Regarding the SSH stuff, I'm going to say that should be a different issue. The whole setup process for SSH channels should be changing soon (see #122 and #91), and that exception doesn't actually break anything. It's just not a pleasant message for an actual exceptional condition. Plus this PR is getting pretty hefty as it is. lol

**Comment by Mitul16** [comment]
Yeah, I should have created a separate __Issue__ here or posted the information on exisiting issues (#122 and #91) if applicable

**Comment by calebstewart** [comment]
> Yeah, I should have created a separate **Issue** here or posted the information on exisiting issues (#122 and #91) if applicable

No problem. If/when you are comfortable that the exception issues are fixed, please mark this PR as reviewed, and then I'll merge. This and #140 are the two things I'd like to get merged before dropping v0.4.3. They're big fixes that should be ideally released ASAP. Thanks again for your help on this! These two issues have been huge steps forward in the stability of pwncat. :smile:

**Comment by Mitul16** [comment]
Just a quick question, how are you trying to catch the `ChannelError` given that you have modified the `create_session` method to go hand-in-hand with the __API__?

I went ahead to test my modifications - two new _exceptions_ raised in `./pwncat/channel/connect.py`

```python
48             try:
49                 client = socket.create_connection((host, port))
50             except socket.gaierror:
51                 raise ChannelError(self, "invalid host provided")
52             except ConnectionRefusedError:
53                 raise ChannelError(self, "connection refused, check your port")
```

**Comment by calebstewart** [comment]
> Just a quick question, how are you trying to catch the `ChannelError` given that you have modified the `create_session` method to go hand-in-hand with the **API**?
> 
> I went ahead to test my modifications - two new _exceptions_ raised in `./pwncat/channel/connect.py`
> 
> ```python
> 48             try:
> 49                 client = socket.create_connection((host, port))
> 50             except socket.gaierror:
> 51                 raise ChannelError(self, "invalid host provided")
> 52             except ConnectionRefusedError:
> 53                 raise ChannelError(self, "connection refused, check your port")
> ```

Whatever calls `create_session` should be prepared to catch any exceptions it raises. It can raise `ChannelError` or `PlatformError` exceptions. In the standard entrypoint, these exceptions are caught at `__main__.py:278` and a "connection failed" message is printed appropriately. 

For the `connect` command used from the local prompt, you are correct. They are not being caught. I'll update that now.

**Comment by Mitul16** [comment]
Yes, because it is being executed from `./pwncat/commands/__init__.py` instead

File: CHANGELOG.md [hunk 1]
```diff
@@ -14,6 +14,8 @@ and simply didn't have the time to go back and retroactively create one.
 ### Changed
 - Changed session tracking so session IDs aren't reused
 - Changed zsh prompt to match CWD of other shell prompts
+- Improved exception handling in `Manager.interactive` ([#133](https://github.com/calebstewart/pwncat/issues/133))
```
Comment (Mitul16) [inline CHANGELOG.md:17]: This need to be updated, the _fixes_ are no longer constrained to `Manager.interactive`

**Review by Mitul16 — DISMISSED** [review]
Ok, I have reviewed all the changes except `pwncat/manager.py`.
There are a lot of modifications and I have not tested each and every little change.
Everything looks fine to me, I will try to fuzz `pwncat/manager.py`

File: CHANGELOG.md (hunk 1 above)
Comment (calebstewart) [reply CHANGELOG.md]: Just pushed a fix which changes that message to be more generic. I don't think going into every single fix is productive in the change log. It just notes that we fixed a lot of exceptions/exception handler things...

**Comment by Mitul16** [comment]
> NOTE: I had a temporary fix for `back` command in my `master` branch and I have removed that _commit_.
You may get into issues with local clone of my fork

**Comment by calebstewart** [comment]
Myself and @JohnHammond have been testing this all evening, and think it's ready for merging. I'm going to go ahead and merge this in as I feel like random bug fixes are getting tossed in (e.g. I just fixed a Windows problem that is only loosely related). The pull request is diverging more than I'd like. If there's more problems, feel free to open another issue and we'll work through it.

**Pull request merged**